crates/red_knot_module_resolver/src/typeshed.rs

```diff
@@ -21,9 +37,71 @@ mod tests {
         assert!(functools_module_stub.is_file());
 
         let mut functools_module_stub_source = String::new();
-        functools_module_stub.read_to_string(&mut functools_module_stub_source)?;
+        functools_module_stub
+            .read_to_string(&mut functools_module_stub_source)
+            .unwrap();
 
         assert!(functools_module_stub_source.contains("def update_wrapper("));
-        Ok(())
+    }
+
+    #[test]
+    fn typeshed_vfs_consistent_with_vendored_stubs() {
+        let vendored_typeshed_stubs = VendoredFileSystem::new(TYPESHED_ZIP_BYTES).unwrap();
+
+        let mut empty_iterator = true;
+        for entry in walkdir::WalkDir::new(&*VENDORED_TYPESHED_DIR).min_depth(1) {
+            empty_iterator = false;
+            let entry = entry.unwrap();
+            let absolute_path = entry.path();
+            let file_type = entry.file_type();
+
+            let relative_path = absolute_path.strip_prefix(&*VENDORED_TYPESHED_DIR).unwrap();
+            let vendored_path = VendoredPath::new(relative_path.to_str().unwrap());
+
+            assert!(vendored_typeshed_stubs.exists(vendored_path));
+
+            let vendored_path_kind = vendored_typeshed_stubs
+                .metadata(vendored_path)
+                .unwrap()
+                .kind();
+            assert_eq!(vendored_path_kind.is_directory(), file_type.is_dir());
+        }
```

Comment:
I don't mind the integration tests, but they are a bit hard to understand. For example, I'm still not a 100% if I understand the second test and I would feel terrified if the test starts failing. 

What's nice about the test is that it is exhaustive, but the exhaustiveness adds a lot of complexity (and file iteration isn't guaranteed to always run in the same order). 

I would probably take a shorter route and simply assert that a few selected files exist.

Reply:
Thanks. I would much rather be exhaustive here, I think. I've gone through and added custom error messages to each `unwrap()` and `assert()` call, so that it's clear exactly which invariant is being asserted at each stage, and so that it should be less terrifying to debug a test if one of the assertions starts failing.